### PR TITLE
types: MsgServer, MsgStream, MMRP

### PR DESCRIPTION
**Pull request by stelcheck**
Type information required for user-land usage of the msgServer
API.

### CODE REVIEW

File: types.d.ts [hunk 1]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
 }
 
+/**
+ * MMRP Envelopes are used to encapsulate data to
+ * be sent between MAGE nodes.
+ *
+ * @class MmrpEnvelope
+ */
+declare class MmrpEnvelope {
+    constructor(eventName: string, messages: string | Buffer | string[] | Buffer[], route: string[], returnRoute: string[], flags: any);
```
Comment (ronkorving) [inline types.d.ts]: flags are: string, int, string[] or int[]

File: types.d.ts [hunk 2]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+    /**
+     *
+     *
+     * @param {string} uri
+     * @param {string} clusterId
+     * @param {any} cb
+     *
+     * @memberof MmrpEnvelope
+     */
+    setMessage(uri: string, clusterId: string, callback: Function): void;
```
Comment (ronkorving) [inline types.d.ts]: signature is (type, message)

Something went wrong here.

File: types.d.ts [hunk 3]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+ * MMRP Envelopes are used to encapsulate data to
+ * be sent between MAGE nodes.
+ *
+ * @class MmrpEnvelope
+ */
+declare class MmrpEnvelope {
+    constructor(eventName: string, messages: string | Buffer | string[] | Buffer[], route: string[], returnRoute: string[], flags: any);
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Sets the message part(s) of the envelope

File: types.d.ts [hunk 4]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     * @param {string} uri
+     * @param {string} clusterId
+     * @param {any} cb
+     *
+     * @memberof MmrpEnvelope
+     */
+    setMessage(uri: string, clusterId: string, callback: Function): void;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Appends a message part to the envelope

File: types.d.ts [hunk 5]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @param {(string | Buffer)} message
+     *
+     * @memberof MmrpEnvelope
+     */
+    addMessage(message: string | Buffer): void;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Sets the route that the envelope needs to take to reach its destination

File: types.d.ts [hunk 6]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @param {(string | string[])} route
+     *
+     * @memberof MmrpEnvelope
+     */
+    setRoute(route: string | string[]): void;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Prepends a route to the existing envelope route

File: types.d.ts [hunk 7]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @param {string} route
```
Comment (ronkorving) [inline types.d.ts]: May also be an array of strings

File: types.d.ts [hunk 8]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @memberof MmrpEnvelope
+     */
+    injectRoute(route: string): void;
+
+    /**
+     *
+     *
+     * @param {*} identity
```
Comment (ronkorving) [inline types.d.ts]: Must be string or undefined (to empty the route)

File: types.d.ts [hunk 9]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     *
+     * @param {string} route
+     *
+     * @memberof MmrpEnvelope
+     */
+    injectRoute(route: string): void;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Removes a route part from the start of the route, if the value matches. Or empties the entire route if the given identity is undefined.

File: types.d.ts [hunk 10]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @param {*} identity
+     *
+     * @memberof MmrpEnvelope
+     */
+    consumeRoute(identity: any): void;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Returns true if the route has not been fully consumed, false otherwise.

File: types.d.ts [hunk 11]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @returns {boolean}
+     *
+     * @memberof MmrpEnvelope
+     */
+    routeRemains(): boolean;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: > Returns the final address in the route.

File: types.d.ts [hunk 12]
```diff
@@ -400,6 +400,225 @@ declare class Logger {
...
+     *
+     * @returns {string}
+     *
+     * @memberof MmrpEnvelope
+     */
+    getFinalDestination(): string;
+
+    /**
+     *
```
Comment (ronkorving) [inline types.d.ts]: Returns true if this envelope tracks and contains the sender's route, false otherwise.

File: types.d.ts (hunk 1 above)
Comment (stelcheck) [inline types.d.ts]: Where may I find a description of their meaning?

Comment (stelcheck) [inline types.d.ts]: NM, found it in the API doc; however, what would happen if I specified both NONE and TRACK_ROUTE in an array?

Comment (ronkorving) [inline types.d.ts]: they're bitwise ORed, and NONE is 0, so it would simply be TRACK_ROUTE.

File: types.d.ts (hunk 7 above)
Comment (stelcheck) [inline types.d.ts]: According to the code, it can only be an array of string, it seems.

File: types.d.ts [hunk 13]
```diff
@@ -400,6 +400,256 @@ declare class Logger {
...
+
+    /**
+     * Set a flag value
+     *
+     * @param {*} flag
+     * @returns {boolean}
+     *
+     * @memberof MmrpEnvelope
+     */
+    setFlag(flag: any): boolean;
```
Comment (ronkorving) [inline types.d.ts]: string or int for the flag param?

**Comment by ronkorving** [comment]
LGTM, you can merge after CI 👍

**Pull request merged**